from PyQt5.uic import loadUi
from PyQt5.QtCore import Qt, QRegExp
from PyQt5.QtWidgets import QWidget
from PyQt5.QtGui import QRegExpValidator, QValidator
from views.misc.filedialog import FileDialog
from views.misc.messagebox import MessageBox
from lib.io.network import ping
from lib.io.parse import CWEToLink
from lib.thread.communicate import Communicate
from lib.thread.exploitthread import ExploitThread
from os import getcwd


class BruteForceView(QWidget):
    def __init__(self, parent):
        super(BruteForceView, self).__init__(parent)
        loadUi('./resources/ui/exploits/bruteforce.ui', self)
        self.__exploit = None
        self.__category = "bf"
        self.__parent = parent

        regex = QRegExp("\d{1,6}")
        self.__validator = QRegExpValidator(regex, self.lineEditPort)
        self.lineEditPort.setValidator(self.__validator)

        # Connect buttons
        self.pushButtonExploit.clicked.connect(self.pushButtonExploitClicked)
        self.pushButtonCancel.clicked.connect(self.pushButtonCancelClicked)
        self.checkBoxVulnerable.stateChanged.connect(self.checkBoxVulnerableStateChanged)
        self.labelAuthor.setTextInteractionFlags(Qt.TextSelectableByMouse)
        self.labelDate.setTextInteractionFlags(Qt.TextSelectableByMouse)
        self.labelTarget.setTextInteractionFlags(Qt.TextSelectableByMouse)
        self.labelDescription.setTextInteractionFlags(Qt.TextSelectableByMouse)
        self.labelDescription.setWordWrap(True)
        self.toolButtonUsers.clicked.connect(self.toolButtonUsersClicked)
        self.toolButtonPasswords.clicked.connect(self.toolButtonPasswordsClicked)

        self.labelCWE.setTextFormat(Qt.RichText)
        self.labelCWE.setTextInteractionFlags(Qt.TextBrowserInteraction)
        self.labelCWE.setOpenExternalLinks(True)

        self.labelReferences.setTextFormat(Qt.RichText)
        self.labelReferences.setTextInteractionFlags(Qt.TextBrowserInteraction)
        self.labelReferences.setOpenExternalLinks(True)

        # Set verticalLayout. It's resize the widget
        self.setLayout(self.verticalLayout)

        # Connect signal and thread
        self.__communicate = Communicate()
        self.__exploitThread = ExploitThread(self.__communicate)
        self.__communicate.finishExploit.connect(self.setResultExploit)

        # Button and progressbar
        self.setProgressBarState(0)
        self.pushButtonCancel.setEnabled(False)

    def check(self):
        if ping(self.__exploit.getIP()):
            if not self.isPort() or not self.lineEditUsers.text() or not self.lineEditPasswords.text():
                MessageBox.critical("Error", "You have to complete the information")
                return False
            else:
                return True
        else:
            MessageBox.critical("Error", "Destination Host Unreachable")
            return False

    def checkBoxVulnerableStateChanged(self):
        if self.checkBoxVulnerable.isChecked():
            self.__parent.addExploitSuccess(self.__exploit)
        else:
            self.__parent.removeExploitSuccess(self.__exploit)

    def getCategory(self):
        return self.__category

    def isPort(self):
        if self.lineEditPort.text():
            state = self.__validator.validate(self.lineEditPort.text(), 0)[0]
            return state == QValidator.Acceptable
        else:
            return False

    def pushButtonCancelClicked(self):
        self.__exploitThread.terminate()
        self.setProgressBarState(3)
        self.pushButtonCancel.setEnabled(False)
        self.pushButtonExploit.setEnabled(True)

    def pushButtonExploitClicked(self):
        if self.check():
            self.plainTextEditData.clear()
            try:
                self.setProgressBarState(2)
                self.pushButtonExploit.setEnabled(False)
                self.pushButtonCancel.setEnabled(True)
                data = self.sanitizeOutput()
                self.__exploitThread.setExploit(self.__exploit, data)
                self.__exploitThread.start()
            except Exception as e:
                MessageBox.critical("Error", str(e.args))

    def sanitizeOutput(self):
        out = {"port": int(self.lineEditPort.text()),
               "users": self.lineEditUsers.text(),
               "passwords": self.lineEditPasswords.text()}

        return out

    def setCheckBoxVulnerableChecked(self, checked=True):
        if checked:
            self.checkBoxVulnerable.setCheckState(Qt.Checked)
        else:
            self.checkBoxVulnerable.setCheckState(Qt.Unchecked)

    def setExploit(self, exploit):
        self.__exploit = exploit
        self.labelNameExploit.setText(self.__exploit.getName())
        self.labelAuthor.setText(self.__exploit.getAuthors())
        self.labelDate.setText(self.__exploit.getDate())
        link = CWEToLink(self.__exploit.getCWE())
        self.labelCWE.setText(link)
        self.labelTarget.setText(self.__exploit.getTarget())
        self.labelReferences.setText(self.__exploit.getReferences())
        self.labelDescription.setText(self.__exploit.getDescription())
        self.plainTextEditData.appendPlainText("")
        default = self.__exploit.default()
        if default:
            self.lineEditPort.setText(default["port"])
            if default["users"].startswith("/"):
                default["users"] = default["users"][1:]

            if default["passwords"].startswith("/"):
                default["passwords"] = default["passwords"][1:]

            self.lineEditUsers.setText("{0}/resources/wordlist/{1}".format(getcwd(), default["users"]))
            self.lineEditPasswords.setText("{0}/resources/wordlist/{1}".format(getcwd(), default["passwords"]))

    def setProgressBarState(self, state):
        if state == 0:
            self.progressBar.hide()
            self.progressBar.setRange(0, 1)
            self.progressBar.setValue(0)
        elif state == 1:
            self.progressBar.setFormat("Waiting")
            self.progressBar.setRange(0, 1)
            self.progressBar.setValue(0)
        elif state == 2:
            self.progressBar.setFormat("Exploiting")
            self.progressBar.setRange(0, 0)
            self.progressBar.show()
        elif state == 3:
            self.progressBar.setFormat("Completed")
            self.progressBar.setRange(0, 1)
            self.progressBar.setValue(1)
            self.progressBar.show()

    def setResultExploit(self, result):
        self.plainTextEditData.appendPlainText(result)
        self.setProgressBarState(3)
        self.pushButtonExploit.setEnabled(True)

    def toolButtonUsersClicked(self):
        self.lineEditUsers.setText(FileDialog.getOpenFileName("txt"))

    def toolButtonPasswordsClicked(self):
        self.lineEditPasswords.setText(FileDialog.getOpenFileName("txt"))
